compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -197,6 +235,137 @@ private void performCodeGen(boolean shrink) {
+
+        if (this.packageContext.project().buildOptions().verbose()) {
+            CodeGenOptimizationReportEmitter.emitCodegenOptimizationReport(
+                    this.usedBIRNodeAnalyzer.pkgWiseInvocationData, getOptimizationReportPath(),
+                    packageContext.project().kind());
+        }
+    }
+
+    private Path getOptimizationReportPath() {
+        return this.packageContext.project().sourceRoot;
+    }
+
+    protected Set<PackageID> getOptimizedPackageIDs() {
+        return this.usedBIRNodeAnalyzer.pkgWiseInvocationData.keySet();
+    }
+
+    private void markTestDependenciesForDuplicateBIRGen() {
+        for (int i = pkgResolution.topologicallySortedModuleList().size() - 1; i >= 0; i--) {
+            ModuleContext moduleContext = pkgResolution.topologicallySortedModuleList().get(i);
+            BLangPackage bLangPackage = moduleContext.bLangPackage();
+
+            if (isRootModule(moduleContext) && bLangPackage.hasTestablePackage()) {
+                Set<BPackageSymbol> buildPkgDependencies = new HashSet<>();
+                Set<BPackageSymbol> testablePkgDependencies = new HashSet<>();
+                collectDependencies(bLangPackage.symbol, buildPkgDependencies);
+                collectDependencies(bLangPackage.getTestablePkg().symbol, testablePkgDependencies);
+
+                buildPkgDependencies.stream()
+                        .filter(testablePkgDependencies::contains).filter(this::isNotWhiteListedPkg)
+                        .forEach(pkgSymbol -> {
+                            pkgSymbol.shouldGenerateDuplicateBIR = true;
+                            // Have to use a hashmap because the pkgIds get mutated later
+                            JvmCodeGenUtil.duplicatePkgsMap.put(
+                                    pkgSymbol.pkgID.orgName + pkgSymbol.pkgID.getNameComps().toString(),
+                                    pkgSymbol.pkgID);
+                        });
+                return;
+            }
+        }
+    }
+
+    private boolean isNotWhiteListedPkg(BPackageSymbol pkgSymbol) {
+            for (String pkgName : WHITE_LISTED_PKG_NAMES) {
+            if (pkgSymbol.pkgID.toString().contains(pkgName)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void collectDependencies(BPackageSymbol pkgSymbol, Set<BPackageSymbol> currentDependencies) {
+        pkgSymbol.imports.forEach(dependency -> {
+            if (currentDependencies.add(dependency)) {
+                collectDependencies(dependency, currentDependencies);
+            }
+        });
+    }
+
+    private void optimizeAndCodegen() {
+        // Codegen cannot be done in the inverted order of the topologicallySortedModuleList.
+        // Therefore, we had to move it into another for loop.
+        for (ModuleContext moduleContext : pkgResolution.topologicallySortedModuleList()) {
+            if (shouldOptimize(moduleContext) || hasTests(moduleContext)) {
```

Comment:
what will happen if the package is testable and it contain at least one usage of the white listed modules?

Reply:
Other imports of the package will be optimized (and if they have common dependencies with its test suit, they will be duplicated as well). But white listed imports will never be optimized or duplicated. Hope I understood the question correctly.